sdk/azcore/runtime/datetime/date_time.go

```diff
@@ -0,0 +1,289 @@
+func (r RFC7231) MarshalJSON() ([]byte, error) {
+	return asJSON(r.String()), nil
+}
+
+func (r RFC7231) MarshalText() ([]byte, error) {
+	return []byte(r.String()), nil
+}
+
+func (r *RFC7231) UnmarshalJSON(data []byte) error {
+	return unmarshalJSON(FormatRFC7231, data, r)
+}
+
+func (r *RFC7231) UnmarshalText(data []byte) error {
+	return unmarshalText(FormatRFC7231, data, r)
+}
+
+func (r RFC7231) String() string {
+	return time.Time(r).Format(http.TimeFormat)
+}
+
+// TimeOnly is a [time.Time] where only the time components have a value.
+type TimeOnly time.Time
+
+func (t TimeOnly) MarshalJSON() ([]byte, error) {
+	return asJSON(t.String()), nil
+}
+
+func (t TimeOnly) MarshalText() ([]byte, error) {
+	return []byte(t.String()), nil
+}
+
+func (t *TimeOnly) UnmarshalJSON(data []byte) error {
+	return unmarshalJSON(FormatTimeOnly, data, t)
+}
+
+func (t *TimeOnly) UnmarshalText(data []byte) error {
+	return unmarshalText(FormatTimeOnly, data, t)
+}
+
+func (t TimeOnly) String() string {
+	return time.Time(t).Format(timeOnly)
+}
+
+// Unix is a [time.Time] represented as a Unix time stamp.
+type Unix time.Time
+
+func (u Unix) MarshalJSON() ([]byte, error) {
+	// unix time stamps are sent as a JSON number, so don't quote them
+	return []byte(u.String()), nil
+}
+
+func (u Unix) MarshalText() ([]byte, error) {
+	return []byte(u.String()), nil
+}
+
+func (u *Unix) UnmarshalJSON(data []byte) error {
+	if data == nil || string(data) == "null" {
+		return nil
+	}
+	// Unix time is a JSON number, no quotes to strip off
+	return u.UnmarshalText(data)
+}
+
+func (u *Unix) UnmarshalText(data []byte) error {
+	// this is to handle XML with an empty value, e.g. <SomeTime />
+	if len(data) == 0 {
+		return nil
+	}
+
+	tt, err := parseUnixTime(string(data))
+	if err != nil {
+		return err
+	}
+	*u = Unix(tt)
+	return nil
+}
+
+func (u Unix) String() string {
+	return fmt.Sprintf("%d", time.Time(u).Unix())
+}
+
+// TypeConstraint is a generic type constraint for the supported date/time formats.
+type TypeConstraint interface {
+	DateOnly | RFC1123 | RFC3339 | RFC7231 | TimeOnly | Unix
+}
+
+func asJSON(v string) []byte {
+	return []byte(`"` + v + `"`)
+}
+
+func parse(format, value string) (time.Time, error) {
+	return time.Parse(format, strings.ToUpper(value))
+}
+
+func unmarshalJSON[T TypeConstraint](format Format, data []byte, dest *T) error {
+	if data == nil || string(data) == "null" {
+		return nil
+	}
+	// strip off quotes
+	data = data[1 : len(data)-1]
+	return unmarshal(format, data, dest)
+}
+
+func unmarshalText[T TypeConstraint](format Format, data []byte, dest *T) error {
+	// this is to handle XML with an empty value, e.g. <SomeTime />
+	if len(data) == 0 {
+		return nil
+	}
+	return unmarshal(format, data, dest)
+}
+
+func unmarshal[T TypeConstraint](format Format, data []byte, dest *T) error {
+	p, err := Parse(format, string(data))
+	if err != nil {
+		return err
+	}
+	*dest = T(p)
+	return nil
+}
+
+const timeOnly = "15:04:05.999999999"
```

Comment:
Hmm, `time.TimeOnly` (Go 1.20) omits the sub-second precision. Maybe we don't need it here either?